### PR TITLE
Feat/606 retry backoff policy

### PR DESCRIPTION
**Pull request by coltmcnealy-lh**
Implements #606

### CODE REVIEW

**Review by mijailr — APPROVED** [review]

File: server/src/main/java/io/littlehorse/common/model/getable/core/taskrun/TaskRunModel.java [hunk 1]
```diff
@@ -298,21 +336,94 @@ public void updateTaskResult(ReportTaskRunModel ce) {
...
+
+    private boolean shouldRetry() {
+        // Shouldn't look at nodeRun to decide whether to retry task or not.
+        TaskAttemptModel latest = getLatestAttempt();
+        if (latest == null) {
+            // This really shouldn't happen I think
+            return false;
+        }
+
+        if (latest.getStatus() != TaskStatus.TASK_FAILED && latest.getStatus() != TaskStatus.TASK_TIMEOUT) {
```
Comment (eduwercamacaro) [inline server/src/main/java/io/littlehorse/common/model/getable/core/taskrun/TaskRunModel.java:390]: [Suggestion / Feel free to ignore]:
new method in the `TaskAttemptModel#shouldRetry`  that contains this line here

File: server/src/main/java/io/littlehorse/common/model/getable/core/taskrun/TaskRunModel.java [hunk 2]
```diff
@@ -298,21 +336,94 @@ public void updateTaskResult(ReportTaskRunModel ce) {
...
+        scheduledTask.setSource(taskRunSource);
+        scheduledTask.setTaskDefId(taskDefId);
+        scheduledTask.setTaskRunId(id);
+
+        processorContext.getTaskManager().scheduleTask(scheduledTask);
+    }
+
+    private boolean shouldRetry() {
+        // Shouldn't look at nodeRun to decide whether to retry task or not.
+        TaskAttemptModel latest = getLatestAttempt();
```
Comment (eduwercamacaro) [inline server/src/main/java/io/littlehorse/common/model/getable/core/taskrun/TaskRunModel.java:384]: Is it possible to receive this latest attempt from the method arguments?

File: server/src/main/java/io/littlehorse/common/model/getable/core/taskrun/TaskRunModel.java [hunk 3]
```diff
@@ -298,21 +336,94 @@ public void updateTaskResult(ReportTaskRunModel ce) {
...
+        if (retryPolicyType != RetryPolicyCase.EXPONENTIAL_BACKOFF) {
+            throw new IllegalStateException("Unimplemented retry policy type");
+        }
+
+        long delayMs = exponentialBackoffRetryPolicy
+                .calculateDelayForNextAttempt(attempts.size())
+                .get();
+        Date maturationTime = new Date(System.currentTimeMillis() + delayMs);
+        LHTimer timer = new LHTimer(new CommandModel(new TaskAttemptRetryReadyModel(id), maturationTime));
+        processorContext.getTaskManager().scheduleTimer(timer);
```
Comment (eduwercamacaro) [inline server/src/main/java/io/littlehorse/common/model/getable/core/taskrun/TaskRunModel.java:423]: `public void scheduleTimer(CoreSubcommand commandTimer, delayMs) {}`
This method in the task manager class could be useful:

File: server/src/main/java/io/littlehorse/common/model/getable/core/taskrun/TaskRunModel.java [hunk 4]
```diff
@@ -298,21 +336,94 @@ public void updateTaskResult(ReportTaskRunModel ce) {
             // Tell the WfRun that the TaskRun is done.
             transitionTo(TaskStatus.TASK_SUCCESS);
         } else if (shouldRetry()) {
-            transitionTo(TaskStatus.TASK_SCHEDULED);
-            scheduleAttempt();
+            scheduleRetryAtAppropriateTime();
```
Comment (eduwercamacaro) [inline server/src/main/java/io/littlehorse/common/model/getable/core/taskrun/TaskRunModel.java:339]: This else-if condition can be replaced by a `maybeRetryAtAppropriateTime`.

**Review by eduwercamacaro — APPROVED** [review]

**Pull request merged**